content/blog/2023-12-11-arrays-and-variable-arguments-in-Bril.md

```diff
@@ -0,0 +1,96 @@
+
+```
+# ARGS: 97 108 98 101 114 116
+@main(size: int, array: ptr<int>) {
+  k: int = const 4;
+  zero: int = const 0;
+  five: int = const 5;
+  output: int = call @quickselect array zero five k;
+  print output;
+}
+```
+
+### Arrays in the TypeScript compiler
+
+For this part, I modified the compiler in `ts2bril.ts`. First, I had to change `tsTypeToBril` so that array types are recognized in Bril. I used the TS typechecker to check if the given type is an array type. If so, we map it to the corresponding pointer type in Bril based on what the element type is in Bril.
+
+Then, I changed `emitExpr` to be able to handle array operations. I started by adding a case in the switch statement for array literals, e.g. `{1, 2, 3}`. We evaluate the Bril type of the array, allocate an array in Bril and iterate over the array to store the elements. Next, I added a case for array indexing. To do this, we obtain Bril expressions for the array and the index by calling `emitExpr`, then create a pointer to the indexed element using `ptradd`. Indices are `number`s in TypeScript, but integers in Bril, which is a problem. The implementation currently accepts any `bigint`s as indices, but only `number` literals can be indices. So a programmer can either use `numbers[0n]` or `numbers[0]`, but `numbers[i]` only works if `i` has type `bigint`. The reason for this is that we can try to parse numeric literals as integers, but there is no narrowing operation for variables in Bril currently, and I am not sure if it would make sense for there to be one. This could be extended in the future.
+
+I also changed the existing assignment case to allow assignment to an element in the array. This is similar to evaluating indexed array as an rvalue, mentioned above.
+
+It is the programmer's responsibility to free the arrays, unlike in real TypeScript, because Bril garbage management is manual by default. I overloaded the existing `free` function in `mem.d.ts` so that the programmer can free the array in TypeScript.
+```
+export function free<T>(array: T): void;
+```
+
+The code below 
+```
+const numbers: bigint[] = [18n,26n];
+console.log(numbers[0n])
+free<bigint[]>(numbers);
+```
+gets compiled to 
+```
+@main {
+  v0: int = const 2;
+  v1: ptr<int> = alloc v0;
+  cur: ptr<int> = id v1;
+  v2: int = const 1;
+  v3: int = const 18;
+  store cur v3;
+  cur: ptr<int> = ptradd cur v2;
+  v4: int = const 26;
+  store cur v4;
+  numbers: ptr<int> = id v1;
+  v5: ptr<int> = id numbers;
+  v6: int = const 0;
+  v7: ptr<int> = ptradd v5 v6;
+  v8: int = load v7;
+  print v8;
+  v9: int = const 0;
+  v10: ptr<int> = id numbers;
+  free v10;
+  v11: int = const 0;
+}
+```
+
+As you can see, there is quite a bit of redundant and dead code due to the way expressions are evaluated in the compiler, but it should be easily eliminated with optimizations.
```

Comment:
And in general, the TS-to-Bril compiler already gleefully emits a lot of dead code!